<launch>
  <!-- size of map, change the size inflate x, y, z according to your application -->
  <arg name="map_size_x" value="40.0"/>
  <arg name="map_size_y" value="40.0"/>
  <arg name="map_size_z" value=" 3.0"/>

  <!-- topic of your odometry such as VIO or LIO -->
  <arg name="odom_topic" value="/visual_slam/odom" />

  <!-- Define parameters -->
  <arg name="odometry_topic" default="/visual_slam/odom"/>
  <arg name="cloud_topic" default="/pcl_render_node/cloud"/>
  <arg name="camera_pose_topic" default="/pcl_render_node/camera_pose"/>
  <arg name="depth_topic" default="/pcl_render_node/depth"/>
  <arg name="flight_type" default="1"/>
  <arg name="planning_horizon" default="2.5"/>
  <arg name="point_num" default="5"/>
  <arg name="map_size_x_" default="40.0"/>
  <arg name="map_size_y_" default="40.0"/>
  <arg name="map_size_z_" default="3.0"/>
  <arg name="max_vel" default="2.0"/>
  <arg name="max_acc" default="2.0"/>
  <arg name="cx" default="320"/>
  <arg name="cy" default="240"/>
  <arg name="fx" default="525"/>
  <arg name="fy" default="525"/>
  <!-- Add other necessary arguments here -->

  <!-- Remapping rules -->
  <remap from="/odom_world" to="$(arg odometry_topic)"/>
  <remap from="/grid_map/odom" to="$(arg odometry_topic)"/>
  <remap from="/grid_map/cloud" to="$(arg cloud_topic)"/>
  <remap from="/grid_map/pose" to="$(arg camera_pose_topic)"/>
  <remap from="/grid_map/depth" to="$(arg depth_topic)"/>

  <!-- Planning FSM parameters -->
  <param name="fsm/flight_type" value="$(arg flight_type)" type="int"/>
  <param name="fsm/thresh_replan" value="1.5" type="double"/>
  <param name="fsm/thresh_no_replan" value="2.0" type="double"/>
  <param name="fsm/planning_horizon" value="$(arg planning_horizon)" type="double"/>
  <!-- Add other FSM parameters here -->

  <!-- Grid map parameters -->
  <param name="grid_map/resolution" value="0.1"/>
  <param name="grid_map/map_size_x" value="$(arg map_size_x_)"/>
  <param name="grid_map/map_size_y" value="$(arg map_size_y_)"/>
  <param name="grid_map/map_size_z" value="$(arg map_size_z_)"/>
  <!-- Add other grid map parameters here -->

  <!-- Planner manager parameters -->
  <param name="manager/max_vel" value="$(arg max_vel)" type="double"/>
  <param name="manager/max_acc" value="$(arg max_acc)" type="double"/>
  <!-- Add other manager parameters here -->

  <!-- Trajectory optimization parameters -->
  <param name="optimization/lambda_smooth" value="1.0" type="double"/>
  <!-- Add other optimization parameters here -->

  <!-- BSpline parameters -->
  <param name="bspline/limit_vel" value="$(arg max_vel)" type="double"/>
  <param name="bspline/limit_acc" value="$(arg max_acc)" type="double"/>
  <param name="bspline/limit_ratio" value="1.1" type="double"/>




  <!-- trajectory server -->
  <node pkg="ego_planner" name="traj_server" type="traj_server" output="screen">
    <remap from="/position_cmd" to="planning/pos_cmd"/>

    <remap from="/odom_world" to="$(arg odom_topic)"/>
    <param name="traj_server/time_forward" value="1.0" type="double"/>
  </node>

  <node pkg="waypoint_generator" name="waypoint_generator" type="waypoint_generator" output="screen">
    <remap from="~odom" to="$(arg odom_topic)"/>
    <remap from="~goal" to="/move_base_simple/goal"/>
    <remap from="~traj_start_trigger" to="/traj_start_trigger" />
    <param name="waypoint_type" value="manual-lonely-waypoint"/>
  </node>

  <!-- use simulator -->
  <include file="$(find ego_planner)/launch/simulator.xml">
    <arg name="map_size_x_" value="$(arg map_size_x)"/>
    <arg name="map_size_y_" value="$(arg map_size_y)"/>
    <arg name="map_size_z_" value="$(arg map_size_z)"/>
    <arg name="c_num" value="200"/>
    <arg name="p_num" value="200"/>
    <arg name="min_dist" value="1.2"/>

    <arg name="odometry_topic" value="$(arg odom_topic)" />
  </include>

</launch>